framework/src/problems/ReferenceResidualProblem.C

```diff
@@ -337,25 +347,38 @@ ReferenceResidualProblem::updateReferenceResidual()
   System & s = nonlinear_sys.system();
   auto & as = aux_sys.sys();
 
-  for (unsigned int i = 0; i < _group_resid.size(); ++i)
-  {
-    _group_resid[i] = 0.0;
-    _group_output_resid[i] = 0.0;
-    _group_ref_resid[i] = 0.0;
-  }
+  std::fill(_group_resid.begin(), _group_resid.end(), 0.0);
+  std::fill(_group_output_resid.begin(), _group_output_resid.end(), 0.0);
+  if (_local_norm)
+    std::fill(_group_ref_resid.begin(), _group_ref_resid.end(), 1.0);
+  else
+    std::fill(_group_ref_resid.begin(), _group_ref_resid.end(), 0.0);
 
   for (unsigned int i = 0; i < _soln_vars.size(); ++i)
   {
-    const auto resid =
-        Utility::pow<2>(s.calculate_norm(nonlinear_sys.RHS(), _soln_vars[i], DISCRETE_L2));
+    Real resid = 0.0;
     const auto group = _variable_group_num_index[i];
-    _group_resid[group] += _converge_on_var[i] ? resid : 0;
-    _group_output_resid[group] += resid;
-    if (_reference_vector)
+    if (_local_norm)
     {
-      const auto ref_resid = s.calculate_norm(*_reference_vector, _soln_vars[i], DISCRETE_L2);
-      _group_ref_resid[group] += Utility::pow<2>(ref_resid);
+      mooseAssert(nonlinear_sys.RHS().size() == (*_reference_vector).size(),
+                  "Sizes of nonlinear RHS and reference vector should be the same.");
+      mooseAssert((*_reference_vector).size(), "Reference vector must be provided.");
+      auto div = nonlinear_sys.RHS().clone();
+      *div /= *_reference_vector;
+      resid = Utility::pow<2>(s.calculate_norm(*div, _soln_vars[i], DISCRETE_L2));
```

Comment:
I think this should be `DISCRETE_L_INF`

Reply:
L2 is not completely wrong here. It would give a convergence criterion somewhere between global and "worst point".

Reply:
I guess it depends on what your objective is. My understanding of the objective of this option is that it should strictly enforce the relative convergence criterion at every DOF, and you'd need the infinity norm for that.

Reply:
I'm not sure we even need the tagging system at all

Reply:
I think you would just be getting the largest residual value for the specified solution variable. Is that what you want? Maybe that works for your problem, but I think in general we still would want that absolute value tag.

Reply:
No...I was wrong...that was dumb...

Reply:
I kinda agree that L_INF norm might be better. L2 norm becomes mesh dependent at a certain point, right? What if there were infinate DOF, you'd never hit L2 no matter what. L_INF is physical (all points below a threshold).